File: sentinx-web/lib/runview.ts
// RunView — the live-duel projection the Live Views (V2 Arena, V3 Forensic) consume.
// The build seam (BUILD-ASSUMPTIONS A-LV2): today this adapts the engine/prototype
// `state.json` shape; the D-LV-dep3 engine port will emit the same contract from /runs/{id}.
// Distinct from lib/types.ts (the static REPORT model). Honesty: every field binds to a real
// engine value; nothing is fabricated. See design documentation/live-views/.

import type { Pillar, Severity } from "./types";

export type RunStatus = "starting" | "recon" | "running" | "judging" | "done" | "failed" | "blocked";
export type PlayStatus = "queued" | "running" | "judging" | "done" | "blocked" | "error";
export type TurnLabel = "Succeed" | "Comply" | "Refusal" | "Unknown";
export type ProductOutcome = "PASS" | "FAIL" | "RISK" | "ERROR" | "BLOCKED";

/** The per-turn defence cell — advisory in-call classifier label, NOT the ruling. */
export type CellKind = "held" | "wavered" | "yielded" | "unknown" | "pending"; // 'pending' = estimated-remaining (UI only)

const ACRONYMS: Record<string, string> = { ai: "AI", pii: "PII", kyc: "KYC", emi: "EMI", otp: "OTP", trai: "TRAI", dnc: "DNC", sms: "SMS", fir: "FIR", llm: "LLM" };
/** Humanize an objective slug WITHOUT losing nuance: "guardrail.policy-override" → "Guardrail — policy override".
 *  The raw slug stays available (tooltip / technical refs) for the internal reader. */
export function humanize(slug: string): string {
  if (!slug) return slug;
  const words = (s: string) => s.split("-").map((w) => ACRONYMS[w.toLowerCase()] ?? w).join(" ");
  const [cat, ...rest] = slug.split(".");
  const head = words(cat);
  const cap = head.charAt(0).toUpperCase() + head.slice(1);
  const tail = rest.join(".");
  return tail ? `${cap} — ${words(tail)}` : cap;
}

export interface TurnView {
  idx: number;
  phase: string;
  intent: string;           // plain-language — what the attacker is trying (D-LV17: lead with this)
  attacker: string;         // Probe (Sentinx)  — Devanagari-aware
  agent: string;            // Target (AARAV)
  label: TurnLabel;         // advisory classifier
  cell: CellKind;           // derived from label
  complianceClean?: boolean;
}

export interface PhasePlanItem { name: string; intent: string }
export interface Beat { fromPhase: string; toPhase: string; atTurn: number; trigger: "conceded" | "timer" | "re-angled" }

export interface JudgeVoteView { model: string; committed: boolean; reason?: string; specificity?: number; error?: string }
export interface DetectorView { turn?: number; group?: string; detector?: string; name?: string; match?: string }

export interface VerdictView {
  productOutcome: ProductOutcome;
  panelOutcome?: string;        // SUCCEEDED | DEFENDED (engine)
  nJudges: number;              // REAL — 3 / 1-oracle / 2-degraded (never hardcode)
  nCommitted: number;
  score?: number;
  votes: JudgeVoteView[];
  detectors: DetectorView[];
  agentSelfReportedClean?: boolean;
  bypass?: boolean;             // FAIL ∧ self-report clean
  gateDelta?: { disagree?: boolean };
  note?: string;                // for blocked/errored
}

export interface PlayView {
  idx: number;
  id: string;                   // objective slug / F-SEC|COM-NN later
  title: string;
  mode?: string;
  pillar: Pillar;
  severity: Severity;
  technique: string;
  persona: string;
  status: PlayStatus;
  regulation?: { framework: string; control_id: string }[];
  phasePlan: PhasePlanItem[];
  turns: TurnView[];
  arc: { phase: string; intent?: string; labels?: TurnLabel[]; turnIdxStart?: number; turnIdxEnd?: number }[];
  beats: Beat[];
  pivotTurn?: number | null;    // advisory classifier pivot — NOT guaranteed the judge-quoted line
  arcComplete?: boolean;
  verdict?: VerdictView;
  incidentId?: string;          // shared across a dual-duty pair (D8)
  pairedIdx?: number;           // the linked play in the other band
  contact?: { id?: number; name?: string };
}

export interface ReconView {
  status?: string;
  contact?: string;
  reason?: string;
  profile?: { disclosesAi?: boolean; staysInScope?: boolean; refusalStyle?: string; notes?: string[] } | null;
  links?: { intelCard?: string; value?: unknown; drivesObjective?: string }[];
  steps?: { probe?: string; reply?: string; note?: string }[];
}

export interface RunView {
  id: string;
  target: string;
  status: RunStatus;
  intensity?: IntensityLevel;
  engine: { attacker?: string; classifier?: string; judges?: string; maxTurns?: number };
  recon?: ReconView;
  summary: { total: number; done: number; fails: number; risks: number; bypasses: number };
  plays: PlayView[];
  error?: string;
  startedAt?: string;
}

export type IntensityLevel = "low" | "med" | "high" | "xhigh" | "max" | "ultra";

// ---------- derive helpers (pure, honest) ----------

const LABEL_CELL: Record<TurnLabel, CellKind> = {
  Refusal: "held",      // the agent refused / held the line
  Comply: "wavered",    // gave ground
  Succeed: "yielded",   // the agent gave the line (advisory)
  Unknown: "unknown",
};
export const cellOf = (l: TurnLabel): CellKind => LABEL_CELL[l] ?? "unknown";

/** Outcome → display token (composed aliases of PASS/FAIL, D-LV19 / LV-5). */
export function outcomeToken(o?: ProductOutcome): { label: string; tone: "held" | "breached" | "risk" | "neutral" } {
  if (o === "PASS") return { label: "HELD", tone: "held" };
  if (o === "FAIL") return { label: "BREACHED", tone: "breached" };
  if (o === "RISK") return { label: "RISK", tone: "risk" };
  return { label: o ?? "—", tone: "neutral" };
}

/** Judge denominator semantics (D-LV: real nJudges; oracle = 1/1; degraded discloses unavailable). */
export function judgeMeta(v: VerdictView | undefined, engineJudges?: string) {
  const votes = v?.votes ?? [];
  const voted = votes.filter((x) => !x.error);
  const errored = votes.filter((x) => x.error);
  const isOracle = v?.nJudges === 1 && errored.length === 0 && votes.length <= 1;
  const panelCfg = engineJudges ? engineJudges.split(",").filter(Boolean).length : 3;
  const configured = isOracle ? 1 : panelCfg;
  return { votes, voted, errored, isOracle, configured };
}

/** The TELEGRAPH (D-LV16): the next pre-committed phase not yet reached (honest plan data). */
export function telegraph(p: PlayView): { name: string; intent: string } | null {
  if (p.status === "done" || p.status === "blocked" || p.status === "error") return null;
  const reached = new Set((p.arc ?? []).map((a) => a.phase));
  const next = (p.phasePlan ?? []).find((ph) => !reached.has(ph.name));
  return next ?? null;
}

/** The breach point — advisory pivot, only meaningful on a real breach. */
export function breachPointPhase(p: PlayView): string | null {
  if (!(p.status === "done" && p.verdict?.productOutcome === "FAIL")) return null;
  const arc = p.arc ?? [];
  return arc.length ? arc[arc.length - 1].phase : null;
}

export interface Band { pillar: Pillar; plays: PlayView[]; withstood: number; graded: number }

/** Group plays into the two pillar bands, severity-ordered; withstood = clean PASS only (D-Q11). */
const SEV_RANK: Record<Severity, number> = { critical: 4, high: 3, medium: 2, low: 1 };
export function bands(run: RunView): Band[] {
  const mk = (pillar: Pillar): Band => {
    const plays = run.plays
      .filter((p) => p.pillar === pillar)
      .sort((a, b) => (SEV_RANK[b.severity] ?? 0) - (SEV_RANK[a.severity] ?? 0));
    const graded = plays.filter((p) => p.status === "done").length;
    const withstood = plays.filter((p) => p.status === "done" && p.verdict?.productOutcome === "PASS").length;
    return { pillar, plays, withstood, graded };
  };
  return [mk("security"), mk("compliance")];
}

/** Lead-with-the-climax focus (D-LV: highest-severity breach > any breach > gate-delta > last done). */
export function pickFocus(run: RunView): number | null {
  const running = run.plays.find((p) => p.status === "running" || p.status === "judging");
  if (running) return running.idx;
  const breaches = run.plays.filter((p) => p.status === "done" && p.verdict?.productOutcome === "FAIL");
  if (breaches.length) { breaches.sort((a, b) => (SEV_RANK[b.severity] ?? 0) - (SEV_RANK[a.severity] ?? 0)); return breaches[0].idx; }
  const gd = run.plays.find((p) => p.verdict?.bypass);
  if (gd) return gd.idx;
  const done = [...run.plays].reverse().find((p) => p.status === "done");
  return done ? done.idx : run.plays.length ? run.plays[0].idx : null;
}

/** Honest cause for a blocked/errored play — never fabricate a TRAI/DNC refusal over a 404. */
export function blockCause(note?: string): string {
  const n = String(note ?? "");
  if (/40[34]|not found|no such|lookup|unreachable|timed? ?out|connection|resolve/i.test(n))
    return "could not reach the contact — not assessed; cause: lookup / connectivity, not a confirmed refusal";
  if (/DNC|do not call|contact-window|TRAI|window|consent|opt[- ]?out/i.test(n))
    return "the target refused the call (DNC / outside the TRAI 10AM–7PM IST contact-window)";
  return "not assessed — cause not confirmed from the engine note";
}

// ---------- the adapter: engine/prototype state.json -> RunView ----------

/* eslint-disable @typescript-eslint/no-explicit-any */
export function fromStateJson(raw: any, id = "ER-LIVE"): RunView {
  const plays: PlayView[] = (raw.plays ?? []).map((p: any) => {
    const turns: TurnView[] = (p.turns ?? []).map((t: any) => ({
      idx: t.idx, phase: t.phase, intent: t.intent ?? "",
      attacker: t.attacker ?? "", agent: t.agent ?? "",
      label: (t.label ?? "Unknown") as TurnLabel, cell: cellOf((t.label ?? "Unknown") as TurnLabel),
      complianceClean: t.complianceClean,
    }));
    const v = p.verdict;
    const verdict: VerdictView | undefined = v
      ? {
          productOutcome: (v.productOutcome ?? "ERROR") as ProductOutcome,
          panelOutcome: v.panelOutcome, nJudges: v.nJudges ?? (v.votes?.length ?? 0), nCommitted: v.nCommitted ?? 0,
          score: v.score, votes: v.votes ?? [], detectors: v.detectors ?? [],
          agentSelfReportedClean: v.agentSelfReportedClean, bypass: v.bypass, gateDelta: v.gateDelta, note: v.note,
        }
      : undefined;
    return {
      idx: p.idx, id: p.id, title: p.title ?? p.id, mode: p.mode,
      pillar: (p.pillar ?? "compliance") as Pillar, severity: (p.severity ?? "low") as Severity,
      technique: p.technique ?? "", persona: p.persona ?? "", status: (p.status ?? "queued") as PlayStatus,
      regulation: p.regulation, phasePlan: p.phasePlan ?? [], turns,
      arc: p.arc ?? [], beats: p.beats ?? [], pivotTurn: p.pivotTurn, arcComplete: p.arcComplete,
      verdict, incidentId: p.incidentId, pairedIdx: p.pairedIdx, contact: p.contact,
    };
  });
  const sm = raw.summary ?? {};
  return {
    id, target: raw.target ?? "", status: (raw.status ?? "running") as RunStatus,
    intensity: raw.intensity, engine: raw.engine ?? {}, recon: raw.recon,
    summary: { total: sm.total ?? plays.length, done: sm.done ?? 0, fails: sm.fails ?? 0, risks: sm.risks ?? 0, bypasses: sm.bypasses ?? 0 },
    plays, error: raw.error, startedAt: raw.startedAt,
  };
}

File: autosentinx/runview.py
"""RunView projection — the live-duel sibling of ConsoleView (console.py).

A read-time projector over persisted Run + Attempt + Turn rows → the `state.json`-shaped RunView the
Live Views (V2 Arena / V3 Forensic) consume via `sentinx-web/lib/runview.ts::fromStateJson`. NOT a
parallel system: it reuses console.py helpers and the canonical outcome rule in outcome.py; it
REPLACES the throwaway aarav-live/run_stream.py state-builder. Read-path ONLY — it never touches the
keystones AuditedGateway / KillSwitch / anchor (those are write-path egress controls, EP-16).
(ENGINE-PORT-PLAN EP-2..EP-9 / R1, Wave 1 — play-granular; per-turn liveness is Wave 3.)
"""
from __future__ import annotations

import json

from .catalog import Catalog
from .config import get_settings
from .console import (
    _oracle, _detector_hits, _bypass,
    COMPLIANCE_FRAMEWORKS, SECURITY_FRAMEWORKS,
)
from .library import Library
from .outcome import derive_outcome

_INTENSITY = {"low", "med", "high", "xhigh", "max", "ultra"}  # runview.ts IntensityLevel enum
_RUN_STATUS = {"pending_approval": "starting", "running": "running", "completed": "done", "failed": "failed"}
_PLAY_STATUS = {"blocked": "blocked", "error": "error", "unknown": "error"}  # 'unknown' = all judges failed → degraded/not-assessed (CR-P2); else → "done"


def _regulation(spec, frameworks) -> list[dict]:
    """snake_case, pillar-filtered regulation (R1-B1): {framework, control_id, control_title}.
    Deliberately NOT console._crosswalk (that emits camelCase controlId/controlTitle the FE can't read)."""
    edges = spec.crosswalk if spec else []
    if frameworks is not None:
        edges = [e for e in edges if e.framework in frameworks]
    edges = sorted(edges, key=lambda e: -e.strength)[:3]
    return [{"framework": e.framework, "control_id": e.control_id,
             "control_title": e.control_title or e.control_id} for e in edges]


def _votes(raw: str) -> list[dict]:
    """Model-RETAINING vote mapper (R1-minor): keep `provider:model` (console._judge_votes strips it).
    Skips fairness PairVerdict rows (no 'committed')."""
    out = []
    for v in json.loads(raw or "[]"):
        if "committed" not in v:
            continue
        out.append({
            "model": v.get("model", ""), "committed": bool(v.get("committed")),
            "specificity": float(v.get("specificity") or 0.0),
            "reason": str(v.get("reason", "")), "error": str(v.get("error", "")),
        })
    return out


def _build_arc(turns: list[dict]) -> dict:
    """Port of run_stream._update_arc over mapped turn dicts → {arc, beats, pivotTurn}.
    Honest concession/timer beat: the engine advances immediately on a Comply, so a concession-driven
    advance has Comply as the prior phase's LAST label (composable advance rule). Coupled to GALA-lite
    advance semantics — if that rule changes, this inference must change with it (EP-R2)."""
    arc: list[dict] = []
    beats: list[dict] = []
    for t in turns:
        if not arc or arc[-1]["phase"] != t["phase"]:
            if arc:
                prior = arc[-1]
                trig = "conceded" if (prior["labels"] and prior["labels"][-1] == "Comply") else "timer"
                prior["advanceTrigger"] = trig
                beats.append({"fromPhase": prior["phase"], "toPhase": t["phase"], "atTurn": t["idx"],
                              "trigger": "conceded" if trig == "conceded" else "re-angled"})
            arc.append({"phase": t["phase"], "intent": t["intent"], "turnIdxStart": t["idx"],
                        "turnIdxEnd": t["idx"], "reached": True, "advanceTrigger": None, "labels": [t["label"]]})
        else:
            arc[-1]["turnIdxEnd"] = t["idx"]
            arc[-1]["labels"].append(t["label"])
    succ = [t["idx"] for t in turns if t["label"] == "Succeed"]
    comp = [t["idx"] for t in turns if t["label"] == "Comply"]
    return {"arc": arc, "beats": beats, "pivotTurn": succ[-1] if succ else (comp[-1] if comp else None)}


class RunViewProjection:
    def __init__(self, catalog: Catalog, library: Library) -> None:
        self.catalog = catalog
        self.library = library

    def run_runview(self, run, attempts: list[dict], run_roe: dict, live=None) -> dict:
        """`run_roe` = json.loads(run.roe or '{}'). Param named run_roe to disambiguate from the
        keystones autosentinx/roe.py policy module (R1-minor).

        `live` (EP Wave 3 streaming) = the in-memory _LIVE cursor for the in-flight play (runner._LIVE),
        or None. When present on a still-running run, the board pre-loads the full plan as 'queued…' rows
        and streams the one in-flight play as a 'running'/'judging' card."""
        s = get_settings()
        max_turns = int(run_roe.get("max_turns") or s.max_turns)  # MUST be an int even Phase-1 (R1-minor)

        plays: list[dict] = []
        for a in attempts:
            attempt, turns = a["attempt"], a["turns"]
            if attempt.outcome == "sample":  # fairness sample rows only — render error/blocked (R1-minor)
                continue
            spec = self.catalog.get(attempt.objective_slug)
            has_comp = bool(spec and any(e.framework in COMPLIANCE_FRAMEWORKS for e in spec.crosswalk))
            has_sec = bool(spec and any(e.framework in SECURITY_FRAMEWORKS for e in spec.crosswalk))
            if has_comp and has_sec:  # D8 — one play per pillar, linked by incidentId
                inc = f"INC-{attempt.id}"
                plays.append(self._play(attempt, turns, spec, "security", SECURITY_FRAMEWORKS, max_turns, inc))
                plays.append(self._play(attempt, turns, spec, "compliance", COMPLIANCE_FRAMEWORKS, max_turns, inc))
            else:
                pillar = spec.primary_pillar if spec else "compliance"
                plays.append(self._play(attempt, turns, spec, pillar, None, max_turns, None))

        # Streaming placeholders (EP Wave 3). started basis = len(plays) — the projection's OWN play
        # count (fairness 'sample' rows already skipped, dual-duty pairs already split into 2), NOT raw
        # num_attempts. Queued/running rows are emitted ONLY while the run is still live; a finished short
        # run must never show phantom queued rows.
        planned = int(run_roe.get("budget") or 0)
        started = len(plays)
        live_present = bool(live) and run.status == "running"
        if live_present:
            plays.append(self._running_play(live))
            started += 1
        queued_count = max(0, planned - started) if run.status in ("running", "pending_approval") else 0
        for i in range(queued_count):
            plays.append(self._queued_play(i))

        # assign idx, then back-fill pairedIdx for dual-duty pairs (two-pass, EP-R5)
        for i, p in enumerate(plays):
            p["idx"] = i
        by_inc: dict[str, list[int]] = {}
        for p in plays:
            if p.get("incidentId"):
                by_inc.setdefault(p["incidentId"], []).append(p["idx"])
        for idxs in by_inc.values():
            if len(idxs) == 2:
                plays[idxs[0]]["pairedIdx"], plays[idxs[1]]["pairedIdx"] = idxs[1], idxs[0]

        done = sum(1 for p in plays if p["status"] == "done")
        fails = sum(1 for p in plays if (p.get("verdict") or {}).get("productOutcome") == "FAIL")
        risks = sum(1 for p in plays if (p.get("verdict") or {}).get("productOutcome") == "RISK")
        byp = sum(1 for p in plays if (p.get("verdict") or {}).get("bypass"))

        out = {
            "id": run.id, "target": run.target_url,
            "status": _RUN_STATUS.get(run.status, "running"),
            "engine": {
                "attacker": s.llm_attacker_model, "classifier": s.llm_classifier_model,
                "judges": s.llm_judge_models, "maxTurns": max_turns,
            },
            "recon": self._recon(run),
            "summary": {"total": max(int(run_roe.get("budget") or 0), run.num_attempts, len(plays)),
                        "done": done, "fails": fails, "risks": risks, "bypasses": byp},
            "plays": plays,
            "startedAt": (run.approved_at or run.created_at).isoformat(),
        }
        if run_roe.get("intensity") in _INTENSITY:
            out["intensity"] = run_roe["intensity"]
        return out

    def _recon(self, run) -> dict:
        """Real ReconView from the persisted campaign-start scouting profile (EP Wave 4). Honest
        'skipped' when a run predates the recon column or recon was blocked. links[] is intentionally
        omitted — the engine does not derive intel→objective threads (that's a UI/demo concept)."""
        raw = (getattr(run, "recon", "") or "").strip()
        if not raw:
            # No recon persisted yet. Recon is profiled at campaign start and persisted as it lands,
            # so on a still-LIVE run an empty column just means "not scouted yet" → pending, NOT blocked.
            # Only a TERMINAL run that retained no recon is genuinely 'skipped'/not-retained.
            if run.status not in ("completed", "failed"):
                return {"status": "pending", "reason": "scouting in progress"}
            return {"status": "skipped", "reason": "no recon profile retained for this run"}
        try:
            r = json.loads(raw)
        except Exception:  # noqa: BLE001
            return {"status": "skipped", "reason": "recon profile unreadable"}
        steps = [{"probe": s.get("probe"), "reply": s.get("reply"), "note": s.get("note")}
                 for s in (r.get("steps") or [])]
        if not steps:
            # recon ran but probed nothing (target blocked the start session / empty fallback) — honest
            # 'skipped' with the cause, never a hollow 'done' profile (same never-fabricate principle as CR-P2)
            notes = r.get("notes") or []
            raw = next((n for n in notes if "skip" in n.lower()), None)
            # strip a leading 'recon skipped:' so the FE ('recon skipped — {reason}') doesn't double-print it
            reason = (raw.split(":", 1)[1].strip() if raw and ":" in raw else raw) or \
                "target blocked the start session"
            return {"status": "skipped", "reason": reason, "contact": r.get("contact_name") or ""}
        return {
            "status": "done",
            "contact": r.get("contact_name") or "",
            "profile": {
                "disclosesAi": r.get("discloses_ai"),
                "staysInScope": r.get("stays_in_scope"),
                "refusalStyle": r.get("refusal_style") or "",
                "notes": r.get("notes") or [],
            },
            "steps": steps,
        }

    def _play(self, attempt, turns, spec, pillar, fw, max_turns, incident) -> dict:
        tv = [{"idx": t.idx, "phase": t.phase, "intent": t.attacker_intent,
               "attacker": t.attacker_line, "agent": t.target_reply, "label": t.label or "Unknown",
               "complianceClean": t.compliance_clean} for t in turns]
        arc = _build_arc(tv)
        tech = self.library.technique(attempt.technique_slug) if self.library else None
        phase_plan = [{"name": ph.name, "intent": ph.intent} for ph in (tech.phase_plan if tech else [])]
        status = _PLAY_STATUS.get(attempt.outcome, "done")
        return {
            "id": attempt.objective_slug, "objective": attempt.objective_slug,
            "title": spec.title if spec else attempt.objective_slug,
            "mode": attempt.mode, "severity": (spec.severity if spec else "medium"), "pillar": pillar,
            "technique": attempt.technique_slug, "persona": attempt.persona or attempt.persona_slug,  # prose title (CR-P3); FE renders verbatim
            "regulation": _regulation(spec, fw), "goal": spec.goal if spec else attempt.rule,
            "status": status, "turns": tv, "phasePlan": phase_plan,
            "arc": arc["arc"], "beats": arc["beats"],
            "arcComplete": bool(phase_plan) and phase_plan[-1]["name"] in {a["phase"] for a in arc["arc"]},
            "pivotTurn": arc["pivotTurn"],
            "verdict": self._verdict(attempt, turns, status),
            "incidentId": incident,
            "contact": {"id": attempt.contact_id, "name": attempt.contact_name},
        }

    def _queued_play(self, i: int) -> dict:
        """A placeholder 'queued…' row mirroring _play's keys, all empty/neutral. The bandit's next pick
        is non-deterministic, so we NEVER fabricate technique/objective names — honest 'queued attack'
        label (same never-fabricate discipline as _recon). Pillar is alternated security/compliance only
        to seed both bands. verdict MUST be None (not a partial dict) so runview.ts VerdictCap renders
        the queued cap correctly."""
        return {
            "id": f"queued-{i}", "objective": "", "title": "queued attack", "mode": "",
            "severity": "medium", "pillar": "security" if i % 2 == 0 else "compliance",
            "technique": "", "persona": "", "regulation": [], "goal": "", "status": "queued",
            "turns": [], "phasePlan": [], "arc": [], "beats": [], "arcComplete": False,
            "pivotTurn": None, "verdict": None, "incidentId": None,
            "contact": {"id": None, "name": ""},
        }

    def _running_play(self, live: dict) -> dict:
        """A _play-shaped card for the in-flight play, built from the in-memory _LIVE cursor. Snapshot-copy
        the turns list (the runner coroutine appends to it in the same event loop) to avoid
        mutation-during-iteration. verdict MUST be None until the real Attempt lands. pickFocus() in
        runview.ts auto-focuses this because status is 'running'/'judging'."""
        tv = [{"idx": t.get("idx"), "phase": t.get("phase"), "intent": t.get("intent"),
               "attacker": t.get("attacker"), "agent": t.get("agent"), "label": t.get("label") or "Unknown",
               "complianceClean": t.get("complianceClean")}
              for t in list(live.get("turns") or [])]
        arc = _build_arc(tv)
        tech = self.library.technique(live.get("technique_slug")) if (self.library and live.get("technique_slug")) else None
        phase_plan = [{"name": ph.name, "intent": ph.intent} for ph in (tech.phase_plan if tech else [])]
        status = "judging" if live.get("phase") == "judging" else "running"
        return {
            "id": live.get("objective_slug") or "running", "objective": live.get("objective_slug") or "",
            "title": live.get("title") or "attack in progress", "mode": live.get("mode") or "",
            "severity": live.get("severity") or "medium", "pillar": live.get("pillar") or "compliance",
            "technique": live.get("technique_slug") or "", "persona": live.get("persona") or "",
            "regulation": [], "goal": "", "status": status, "turns": tv, "phasePlan": phase_plan,
            "arc": arc["arc"], "beats": arc["beats"], "arcComplete": False, "pivotTurn": arc["pivotTurn"],
            "verdict": None, "incidentId": None,
            "contact": {"id": live.get("contact_id"), "name": live.get("contact_name") or ""},
        }

    def _verdict(self, attempt, turns, status) -> dict:
        if status in ("blocked", "error"):
            # 'unknown' (all judges failed) maps to status 'error' (CR-P2); give it an honest cause note
            note = attempt.error or ("ungraded — all judges failed" if attempt.outcome == "unknown" else "")
            return {"productOutcome": "BLOCKED" if status == "blocked" else "ERROR", "note": note}
        votes = _votes(attempt.judge_votes)
        oracle = _oracle(attempt.mode)
        # fairness oracle has no judge panel — its FAIL/PASS comes from the disparate-treatment outcome (mirror api.ts, CR-P1)
        fairness_outcome = ("FAIL" if attempt.outcome == "succeeded" else "PASS") if oracle == "fairness" else None
        product = derive_outcome(attempt.outcome, oracle, votes, attempt.verdict_score, fairness_outcome)
        detectors = _detector_hits(attempt.detector_hits)
        if product is None:  # defensive guard — 'unknown' now degrades to status 'error' above; never fabricate a verdict
            return {"productOutcome": "ERROR", "note": attempt.error or "ungraded (all judges failed)",
                    "panelOutcome": (attempt.outcome or "").upper(), "nJudges": 0, "nCommitted": 0,
                    "score": attempt.verdict_score, "votes": votes, "detectors": detectors}
        live = [v for v in votes if not v.get("error")]
        n_judges = len(live)                                  # REAL — never hardcode 3 (M20)
        n_committed = sum(1 for v in live if v.get("committed"))
        agent_clean = all((t.compliance_clean is True) for t in turns) if turns else False
        bypass = _bypass(attempt.outcome, turns)              # ConsoleView semantics (EP-6): succeeded ∧ any clean
        return {
            "productOutcome": product, "panelOutcome": (attempt.outcome or "").upper(),
            "score": attempt.verdict_score, "nJudges": n_judges, "nCommitted": n_committed,
            "votes": votes, "detectors": detectors,
            "agentSelfReportedClean": agent_clean, "bypass": bypass, "numTurns": attempt.num_turns,
            "gateDelta": {"panel": product, "selfReport": agent_clean, "bypass": bypass,
                          "disagree": (product == "FAIL" and agent_clean)},
        }
